radish/extensions/logfile_writer.py

```diff
@@ -0,0 +1,106 @@
+# -*- coding: utf-8 -*-
+
+"""
+    This module provides an extension to write all features, scenarios and steps to a log file.
+"""
+
+import os
+from radish.terrain import world
+from radish.feature import Feature
+from radish.hookregistry import before, after
+from radish.extensionregistry import extension
+
+
+@extension
+class LogFileWriter(object):
+    """
+        Extension to write a log file. It is supported by every platform.
+        It will write the same content as the syslog extension but into a
+        file.
+    """
+
+    OPTIONS = [("--log=<logfile>", "log all of your features, scenarios, and steps to a log file")]
+    LOAD_IF = staticmethod(lambda config: config.log)
+    LOAD_PRIORITY = 80
+
+    def __init__(self):
+        before.all(self.logfile_writer_before_all)
+        before.each_feature(self.logfile_writer_before_each_feature)
+        before.each_scenario(self.logfile_writer_before_each_scenario)
+        before.each_step(self.logfile_writer_before_each_step)
+        after.all(self.logfile_writer_after_all)
+        after.each_feature(self.logfile_writer_after_each_feature)
+        after.each_scenario(self.logfile_writer_after_each_scenario)
+        after.each_step(self.logfile_writer_after_each_step)
+
+    def get_scenario_feature(self, scenario):
+        """
+            Gets the scenarios feature
+        """
+        if not isinstance(scenario.parent, Feature):
+            return scenario.parent.parent
+
+        return scenario.parent
+
+    def log(self, message):
+        """
+            Logs the given message to a log file
+
+            :param string message: the message to log
+        """
+        with open(world.config.log, "a+") as f:
```

Comment:
Yeah, me too. 

I'm not sure if the current implementation really does what we want to achieve here - or at least my understanding of why this feature is of any use.

My understanding of the use-case is that the user might want to add markers of the radish test run to the same log file which is also used by the software-under-test. With that you would exactly see in the log file which logs from the SUT happened in which Steps of the test run.
So, you might end up with something like this:

```
Mar 21 18:15:33    host    begin step 1.1.1. Given I have a step
Mar 21 18:15:33    host    SOME CRAZY LOG FROM THE SUT
...
Mar 21 18:15:33    host    passed step 1.1.1. Given I have a step
Mar 21 18:15:33    host    begin step 1.1.2. When I have some other step
Mar 21 18:15:33    host    SOME OTHER CRAZY LOG FROM THE SUT
...
Mar 21 18:15:33    host    THE SUT FAILED BECAUSE OF ...
Mar 21 18:15:33    host    failed step 1.1.2. When I have some other step
...
```

So far so good about the use-case. (That's actually how we are using radish with syslog.)

The problem I see with the current implementation is that we open a file - however, because of the fact how most filesystems work it's not safe to open and write to the same file within multiple processes or even via multiple file descriptors. This would lead to a mess in the log file. It might look like this:


```
Mar 21 18:15:33    host    begin step 1.1.1. Mar 21 18:15:33    host    SOME CRAZY LOG 
Given I have a step
FROM THE SUT
...
Mar 21 18:15:33    host    passed step 1.1.1. Given I Mar 21 18:15:33   
 host    begin step 1.1.2. When I have so
have a step
me other step
Mar 21 18:15:33    host    SOME OTHER CRAZY LOG FROM THE SUT
...
Mar Mar 21 18:15:33    host    failed step 1.1.2. Whe
21 18:15:33    host    THE SUT FAILED BECAUSE OF ...
n I have some other step
...
```

I think you get the problem I see.

Now, if we'd argue that we (radish) are the exclusive owner of this log file and no one else should write to it ... I wouldn't see the point in this feature in the first place because you'd just get the markers - you can get a similar and much better result using the report formats radish provides (bdd-xml, cucumber json) or just `tee`-ing the output:

```bash
radish features/ | tee -a mylogfile.log
```

Another side-note: you don't want to use `__del__` in like 99.99% of the times you think a destructor might be useful. The thing about `__del__` in Python is, that it's defined that the object should be removed once *all references* to it are deleted, **but** it's not specified that this happens right after the reference count is 0 or you called `del`. The Python garbage collector can remove the object whenever it feels like it :hankey: 

What do you think? Maybe you can think about a legit solution?